test/integration/validation/certificate_test.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2022 The cert-manager Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package validation
+
+import (
+	"context"
+	"strings"
+	"testing"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/cert-manager/cert-manager/pkg/api"
+	cmapi "github.com/cert-manager/cert-manager/pkg/apis/certmanager/v1"
+	cmmeta "github.com/cert-manager/cert-manager/pkg/apis/meta/v1"
+	"github.com/cert-manager/cert-manager/test/integration/framework"
+)
+
+var certificateGVK = schema.GroupVersionKind{
+	Group:   "cert-manager.io",
+	Version: "v1",
+	Kind:    "Certificate",
+}
+
+func TestValidationCertificate(t *testing.T) {
+	tests := map[string]struct {
+		input       runtime.Object
+		errorSuffix string
+		expectError bool
+	}{
+		"Happy path returns no errors": {
+			input: &cmapi.Certificate{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "testing",
+					Namespace: "default",
+				},
+				Spec: cmapi.CertificateSpec{
+					SecretName: "testing-tls",
+					DNSNames:   []string{"myhostname.com"},
+					Usages:     []cmapi.KeyUsage{},
+					IssuerRef: cmmeta.ObjectReference{
+						Name: "letsencrypt-staging",
+					},
+					PrivateKey: &cmapi.CertificatePrivateKey{
+						RotationPolicy: "Always",
```

Comment:
Shall we also add a happy-path test for "Never" or is that overkill?

Reply:
I think this is enough, the other test will actually catch regression on the enum as it expects them in the error string.

We can easily add more test cases for any future regressions though